File: scripts/item-generator/catalog/translations.ts
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Ian Lucas. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

import { stripHtml } from "string-strip-html";
import { type CS2ItemTranslation } from "../../../src/economy-types.ts";
import { assert, ensure } from "../../../src/utils.ts";
import { FORMATTED_STRING_RE } from "../config.ts";
import { type ItemGeneratorContext } from "../types.ts";

function resolveToken(token?: string) {
    return (token?.charAt(0) === "#" ? token.substring(1) : token)?.toLowerCase();
}

function isTranslationKey(ctx: ItemGeneratorContext, token?: string) {
    if (token === undefined || token.length === 0) {
        return false;
    }
    const resolved = resolveToken(token);
    return resolved !== undefined && ensure(ctx.csgoTranslationByLanguage.english)[resolved] !== undefined;
}

export function findTranslation(ctx: ItemGeneratorContext, token?: string, language = "english"): string | undefined {
    token = resolveToken(token);
    if (token === undefined) {
        return undefined;
    }
    const value = ensure(ctx.csgoTranslationByLanguage[language])[token];
    return value !== undefined ? stripHtml(value).result : undefined;
}

export function requireTranslation(ctx: ItemGeneratorContext, token?: string, language = "english"): string {
    return ensure(findTranslation(ctx, token, language), `Failed to find translation for '${token}' (${language}).`);
}

export function hasTranslation(ctx: ItemGeneratorContext, token?: string): boolean {
    token = resolveToken(token);
    return token !== undefined && ensure(ctx.csgoTranslationByLanguage.english)[token] !== undefined;
}

export function addTranslation(
    ctx: ItemGeneratorContext,
    id: number,
    property: keyof CS2ItemTranslation,
    ...tokens: (string | undefined)[]
): void {
    for (const [language, items] of Object.entries(ctx.itemTranslationByLanguage)) {
        const itemLanguage = (items[id] ??= {} as CS2ItemTranslation);
        const value = tokens
            .map((token) => {
                assert(token !== undefined);
                return isTranslationKey(ctx, token)
                    ? (findTranslation(ctx, token, language) ?? requireTranslation(ctx, token))
                    : token;
            })
            .join("")
            .trim();
        if (property === "name" && language === "english" && value.length > 0) {
            ctx.itemNames.set(id, value);
        }
        if (value.length > 0) {
            itemLanguage[property] = value;
        }
    }
}

export function tryAddTranslation(
    ctx: ItemGeneratorContext,
    id: number,
    property: keyof CS2ItemTranslation,
    token: string | undefined
): void {
    if (isTranslationKey(ctx, token)) {
        addTranslation(ctx, id, property, token);
    }
}

export function addFormattedTranslation(
    ctx: ItemGeneratorContext,
    id: number,
    property: keyof CS2ItemTranslation,
    key?: string,
    ...values: string[]
): void {
    for (const [language, items] of Object.entries(ctx.itemTranslationByLanguage)) {
        (items[id] ??= {} as CS2ItemTranslation)[property] = (
            findTranslation(ctx, key, language) ?? requireTranslation(ctx, key, "english")
        ).replace(FORMATTED_STRING_RE, (_, index) => {
            const valueKey = values[parseInt(index, 10) - 1];
            return findTranslation(ctx, valueKey, language) ?? requireTranslation(ctx, valueKey, "english");
        });
    }
}

File: scripts/item-generator/Catalog/Translations.cs
using System.Text.RegularExpressions;
using System.Web;

namespace ItemGenerator.Catalog;

public static class Translations
{
    private static string? ResolveToken(string? token)
    {
        if (token == null || token.Length == 0) return null;
        return (token[0] == '#' ? token[1..] : token).ToLowerInvariant();
    }

    private static bool IsTranslationKey(ItemGeneratorContext ctx, string? token)
    {
        if (token == null || token.Length == 0) return false;
        var resolved = ResolveToken(token);
        if (resolved == null) return false;
        return ctx.CsgoTranslationByLanguage.TryGetValue("english", out var english)
            && english.ContainsKey(resolved);
    }

    public static string? FindTranslation(ItemGeneratorContext ctx, string? token, string language = "english")
    {
        token = ResolveToken(token);
        if (token == null) return null;
        if (!ctx.CsgoTranslationByLanguage.TryGetValue(language, out var translations))
            return null;
        if (!translations.TryGetValue(token, out var value) || value == null)
            return null;
        return StripHtml(value);
    }

    public static string RequireTranslation(ItemGeneratorContext ctx, string? token, string language = "english")
    {
        return FindTranslation(ctx, token, language)
            ?? throw new InvalidOperationException($"Failed to find translation for '{token}' ({language}).");
    }

    public static bool HasTranslation(ItemGeneratorContext ctx, string? token)
    {
        var resolved = ResolveToken(token);
        if (resolved == null) return false;
        return ctx.CsgoTranslationByLanguage.TryGetValue("english", out var english)
            && english.ContainsKey(resolved);
    }

    public static void AddTranslation(ItemGeneratorContext ctx, int id, string property, params string?[] tokens)
    {
        foreach (var (language, items) in ctx.ItemTranslationByLanguage)
        {
            if (!items.TryGetValue(id, out var itemTranslation))
            {
                itemTranslation = new CS2ItemTranslation();
                items[id] = itemTranslation;
            }

            var parts = new List<string>();
            foreach (var token in tokens)
            {
                if (token == null)
                    throw new InvalidOperationException("Translation token is null");
                if (IsTranslationKey(ctx, token))
                    parts.Add(FindTranslation(ctx, token, language) ?? RequireTranslation(ctx, token));
                else
                    parts.Add(token);
            }

            var value = string.Join("", parts).Trim();
            if (value.Length == 0) continue;

            if (property == "name" && language == "english")
                ctx.ItemNames[id] = value;

            SetTranslationProperty(itemTranslation, property, value);
        }
    }

    public static void TryAddTranslation(ItemGeneratorContext ctx, int id, string property, string? token)
    {
        if (IsTranslationKey(ctx, token))
            AddTranslation(ctx, id, property, token!);
    }

    public static void AddFormattedTranslation(ItemGeneratorContext ctx, int id, string property, string? key, params string[] values)
    {
        foreach (var (language, items) in ctx.ItemTranslationByLanguage)
        {
            if (!items.TryGetValue(id, out var itemTranslation))
            {
                itemTranslation = new CS2ItemTranslation();
                items[id] = itemTranslation;
            }

            var template = FindTranslation(ctx, key, language) ?? RequireTranslation(ctx, key, "english");
            var result = Config.FormattedStringRe.Replace(template, match =>
            {
                var index = int.Parse(match.Groups[1].Value) - 1;
                if (index < 0 || index >= values.Length) return match.Value;
                var valueKey = values[index];
                return FindTranslation(ctx, valueKey, language) ?? RequireTranslation(ctx, valueKey, "english");
            });

            SetTranslationProperty(itemTranslation, property, result);
        }
    }

    private static void SetTranslationProperty(CS2ItemTranslation translation, string property, string value)
    {
        switch (property)
        {
            case "name": translation.Name = value; break;
            case "desc": translation.Desc = value; break;
            case "category": translation.Category = value; break;
            case "collectionName": translation.CollectionName = value; break;
            case "collectionDesc": translation.CollectionDesc = value; break;
            case "tournamentDesc": translation.TournamentDesc = value; break;
        }
    }

    private static string StripHtml(string input)
    {
        var result = Regex.Replace(input, "<[^>]+>", "");
        return HttpUtility.HtmlDecode(result) ?? result;
    }
}
